numpy/lib/index_tricks.py

```diff
@@ -146,24 +146,25 @@ def __init__(self, sparse=False):
     def __getitem__(self, key):
         try:
             size = []
-            typ = int
+            # Mimick the behavior of `np.arange` and use a data type
+            # which is at least as large as `int` / `np.long`
+            num_list = [0]
             for k in range(len(key)):
                 step = key[k].step
                 start = key[k].start
+                stop = key[k].stop
                 if start is None:
                     start = 0
                 if step is None:
                     step = 1
                 if isinstance(step, (_nx.complexfloating, complex)):
-                    size.append(int(abs(step)))
-                    typ = float
```

Comment:
With this line removed, what is `np.mgrid[1:2:1j,].dtype`, and does it match `np.mgrid[1:2:1j].dtype?

Reply:
It does, but it doesn't seem to be explicitly tested anywhere. 
I'll add one in a moment.

Examples
---------
``` python
In [1]: import numpy as np

In [2]: np.mgrid[1:2:1j,].dtype
Out[2]: dtype('float64')

In [3]: np.mgrid[1:2:1j].dtype
Out[3]: dtype('float64')
```

Reply:
Oh hang on, I did actually add a test for test:

``` python

        # Related to #16945
        assert_array_equal(
            mgrid[0.1:0.3:3.3j], mgrid[0.1:0.3:3.3j,][0],
        )
```